bin/i18n-map-extractor.js

```diff
@@ -0,0 +1,145 @@
+const recast = require( 'recast' );
+const { forEach, reduce, includes, isEmpty, isFunction } = require( 'lodash' );
+const { readFileSync, writeFileSync } = require( 'fs' );
+const NormalModule = require( 'webpack/lib/NormalModule' );
+
+class wpi18nExtractor {
+	constructor( options ) {
+		const DEFAULT_FUNCTIONS = [
+			'__',
+			'_n',
+			'_x',
+			'_nx',
+			'sprintf',
+		];
+		this.options = options || {};
+		this.options.aliases = this.options.aliases || {};
+		this.options.filename = this.options.filename || 'translation-map.json';
+		this.translationMap = {};
+		this.functionNames = this.options.functionNames || DEFAULT_FUNCTIONS;
+	}
+
+	extractStringFromFunctionCall( args, functionName ) {
+		const strings = [];
+		switch ( functionName ) {
+			case '__':
+			case 'sprintf':
+			case '_n':
+				strings.push( args[ 0 ].value );
+				break;
+			case '_x':
+				strings.push( args[ 1 ].value + '\u0004' + args[ 0 ].value );
+				break;
+			case '_nx':
+				strings.push( args[ 3 ].value + '\u0004' + args[ 0 ].value );
+		}
+		return strings;
+	}
+
+	getStringsFromModule( module, extractor ) {
+		const source = module.originalSource().source();
+		if ( isEmpty( source ) ) {
+			return [];
+		}
+		let strings = [];
+		try {
+			const ast = recast.parse( source );
+			const { types } = recast;
+			types.visit( ast, {
+				visitCallExpression: function( path ) {
+					const node = path.node;
+					if ( includes( extractor.functionNames, node.callee.name ) &&
+						node.arguments
+					) {
+						strings = strings.concat(
+							extractor.extractStringFromFunctionCall(
+								types.getFieldValue( node, 'arguments' ),
+								node.callee.name,
+							)
+						);
+					}
+					this.traverse( path );
+				},
+			} );
+		} catch ( e ) {
+			//we just want to skip parsing errors.
+		}
+		return strings;
+	}
+
+	parseSourcesToMap( modules, chunkName, extractor ) {
+		const { getStringsFromModule } = extractor;
+		reduce(
+			Array.from( modules ),
+			function( mapped, module ) {
+				if ( ! ( module instanceof NormalModule ) ||
+					! isFunction( module.originalSource )
+				) {
+					return mapped;
+				}
+				if ( ! mapped.hasOwnProperty( chunkName ) ) {
+					mapped[ chunkName ] = [];
+				}
+				mapped[ chunkName ] = mapped[ chunkName ]
+					.concat( getStringsFromModule( module, extractor ) );
+				return mapped;
+			},
+			extractor.translationMap
```

Comment:
I _think_ I used `extractor.translationMap` here because it can't be a const with me mutating its value.  I'll have to doublecheck though because `const` in javascript works slightly different than php whereby I think you can't change the data-type on the const but you can still mutate its shape.  So if true, then I should be doing:

```
const { getStringsFromModule, translationMap } = extractor;
```

Reply:
ya that last line makes sense and follows what you were doing with the other vars in terms of assignment.

My understanding of objects assigned as a `const` is that they can't have their assignment mutated (ie: can't swap object A for object B) but you **can** mutate the properties of the object. 
So for example,:

```js
const objectA = { var1 = 1 };
objectA.var1 = 2; // mutating property, not constant, is otay
objectA = { var2 = 2 }; // mutating constant itself is not otay
```